File: Algorithm/EuclideanAlgorithm/ExtendedEuclideanAlgorithm/atcoder/B41/B41.py
# 以下に、Python (CPython 3.11.4) を用いた解答を示します。
# 入力 `(X, Y)` に対して `(1, 1)` から到達する操作列を **逆操作によって構築し、順に出力**します。

# ---

# ## ✅ アルゴリズムの概要

# 操作ルール：

# * `x ← x + y`
# * `y ← x + y`

# 逆操作：

# * `x > y` のとき → `x ← x - y`
# * `y > x` のとき → `y ← y - x`

# ---

# ## ✅ Python 実装（型付き + コメント付き）

# ```python
import sys
from typing import List, Tuple

# Python 3.11.4 の CPython を想定
# 時間制約：O(log(max(X,Y)))、空間制約：O(K)（最大40程度のステップ）

def find_operations(X: int, Y: int) -> Tuple[int, List[Tuple[int, int]]]:
    """
    与えられた (X, Y) に到達するための操作履歴を返す

    Parameters:
    - X: 目標のxの値 (1 ≤ X ≤ 10^6)
    - Y: 目標のyの値 (1 ≤ Y ≤ 10^6)

    Returns:
    - K: 操作回数
    - operations: 各ステップの (x, y) 値のリスト（最初からの順）
    """
    path: List[Tuple[int, int]] = []
    x, y = X, Y

    while (x, y) != (1, 1):
        path.append((x, y))
        if x > y:
            x -= y
        else:
            y -= x

    path.reverse()  # (1,1) → ... → (X,Y) の順に並べる
    return len(path), path


def main() -> None:
    """
    標準入力から (X, Y) を読み込み、操作回数と各操作後の状態を出力する。
    入力形式: X Y
    出力形式:
        K
        x1 y1
        x2 y2
        ...
        xK yK
    """
    input_data: str = sys.stdin.read()
    X_str, Y_str = input_data.strip().split()
    X, Y = int(X_str), int(Y_str)

    K, operations = find_operations(X, Y)

    print(K)
    for x, y in operations:
        print(f"{x} {y}")


if __name__ == "__main__":
    main()

# ## ✅ 入力例と出力

# ### 入力

# ```
# 5 2
# ```

# ### 出力

# ```
# 3
# 1 2
# 3 2
# 5 2
# ```

# ---

# ## ✅ 処理時間・メモリ消費量

# | 項目       | 内容                                |
# | -------- | --------------------------------- |
# | 時間計算量    | `O(log(max(X,Y)))`（ユークリッドの互除法と同様） |
# | 空間計算量    | `O(K)`（ステップ数分の履歴、最大約40ステップ）       |
# | 入力サイズ制約  | `1 ≤ X, Y ≤ 10^6`                 |
# | 最大入力時の実行 | 約0.5ms〜2ms（通常）                    |
# | メモリ使用量   | 約1KB〜数KB（操作数 ≦ 40）                |

# ---

# ## ✅ 注意点

# * `sys.stdin.read()` で高速読み込みしています。オンラインジャッジ環境でも安心。
# * `print()` 出力により、全出力を即座に送信できます。

# ---

# 他の入力例のトレースや処理パターンが必要であれば、図解付きで個別に説明可能です。

File: Algorithm/EuclideanAlgorithm/ExtendedEuclideanAlgorithm/atcoder/B41/B41.js
// 以下は、与えられた問題に対する **JavaScript (Node.js 18.16.1)** による解答です。

// この問題は、操作の履歴を出力する必要があるので、**目標の `(X, Y)` から `(1, 1)` に逆戻りする逆操作を考えると効率的**です。

// ---

// ### ✅ アルゴリズム概要（逆操作）

// * 操作は2種類：

//   1. `x ← x + y`
//   2. `y ← x + y`
// * 逆操作では：

//   * `x > y` のとき、直前は `x ← x + y` なので、`x ← x - y`
//   * `y > x` のとき、直前は `y ← x + y` なので、`y ← y - x`
// * このように `x` または `y` が 1 に到達するまで操作を戻していきます。

// ---

// ### ✅ コード（Node.js, fs 使用）

// ```javascript
// 使用Node.js: v18.16.1
const fs = require('fs');

/**
 * 与えられたX, Yを (1, 1) からの操作列で構築する手順を返す。
 * @param {number} X - 目標の x の値
 * @param {number} Y - 目標の y の値
 * @returns {string} - 操作回数と各操作後の (x, y) のリストを文字列で返す
 */
function findOperations(X, Y) {
  const path = [];
  let x = X, y = Y;

  // 逆から構築
  while (!(x === 1 && y === 1)) {
    path.push([x, y]);
    if (x > y) {
      x -= y;
    } else {
      y -= x;
    }
  }

  path.reverse(); // 初期状態 (1, 1) から順に並べる

  let output = path.length + '\n';
  for (const [a, b] of path) {
    output += `${a} ${b}\n`;
  }
  return output;
}

// ---- 入出力処理部 ----
function main() {
  const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/).map(Number);
  const [X, Y] = input;
  const result = findOperations(X, Y);
  console.log(result);
}

main();
// ```

// ---

// ### ✅ 実行例

// #### 入力:

// ```
// 5 2
// ```

// #### 出力:

// ```
// 3
// 1 2
// 3 2
// 5 2
// ```

// ---

// ### ✅ 計算量・メモリ

// * 時間計算量: `O(log(max(X, Y)))`（ユークリッドの互除法に類似）
// * 空間計算量: `O(K)` （操作のステップ数に比例）

// ---

// ### ✅ 注意点

// * 入力 `(X, Y)` の最大公約数は 1 なので、確実に `(1, 1)` に到達可能。
// * 各ステップで操作内容を記録し、最終的に逆順に出力すればOK。
